File: app/src/main/java/edu/monash/libraryofalexandria/BookAdapter.java
package edu.monash.libraryofalexandria;

import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.ImageView;
import android.widget.TextView;

import java.util.List;
import java.util.Locale;

public class BookAdapter extends RecyclerView.Adapter<BookAdapter.BookViewHolder> {
    private List<Book> bookList;
    private boolean isSearchType;
    private BookItemClickListener listener;

    BookAdapter(BookItemClickListener listener, List<Book> bookList, boolean isSearchType) {
        this.listener = listener;
        this.bookList = bookList;
        this.isSearchType = isSearchType;
    }

    @NonNull
    @Override
    public BookViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        int inflatedResource = isSearchType ? R.layout.item_search_book : R.layout.item_book;
        View view = LayoutInflater.from(parent.getContext()).inflate(inflatedResource, parent, false);
        final BookViewHolder viewHolder = new BookViewHolder(view, isSearchType);
        view.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                listener.onItemClicked(view, viewHolder.getAdapterPosition());
            }
        });
        return viewHolder;
    }

    @Override
    public void onBindViewHolder(@NonNull BookViewHolder holder, int position) {
        holder.setData(bookList.get(position));
    }

    @Override
    public int getItemCount() {
        return bookList == null ? 0 : bookList.size();
    }

    public void updateItems() {
        notifyDataSetChanged();
    }

    static class BookViewHolder extends RecyclerView.ViewHolder implements AdapterView.OnItemClickListener {
        private ImageView bookImageView;
        private TextView bookNameTextView;
        private TextView authorTextView;
        private TextView editionTextView;
        private TextView descriptionTextView;
        private boolean isSearchType;

        BookViewHolder(View itemView, boolean isSearchType) {
            super(itemView);

            this.isSearchType = isSearchType;
            if (isSearchType) {
                descriptionTextView = itemView.findViewById(R.id.book_description_text_view);
            } else {
                bookImageView = itemView.findViewById(R.id.book_image_view);
                editionTextView = itemView.findViewById(R.id.edition_text_view);
            }
            bookNameTextView = itemView.findViewById(R.id.book_name_text_view);
            authorTextView = itemView.findViewById(R.id.book_author_text_view);
        }

        public void setData(Book book) {
            if (isSearchType) {
                descriptionTextView.setText(book.getDescription());
            } else {
                //Set book
                editionTextView.setText(String.format(Locale.ENGLISH, "Edition %d", book.getEdition()));
            }
            bookNameTextView.setText(book.getName());
            authorTextView.setText(book.getAuthor());
        }

        @Override
        public void onItemClick(AdapterView<?> adapterView, View view, int i, long l) {
            Log.e("Selcted Item", "onItemClick: " + i);
        }
    }
}

File: app/src/main/java/edu/anbu/libraryofalexandria/BookAdapter.kt
package edu.anbu.libraryofalexandria

import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.AdapterView
import android.widget.AdapterView.OnItemClickListener
import android.widget.ImageView
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import edu.anbu.libraryofalexandria.BookAdapter.BookViewHolder
import java.util.Locale

class BookAdapter : RecyclerView.Adapter<BookViewHolder> {
    private var bookList: MutableList<Book>? = null
    private var isSearchType: Boolean
    private var listener: BookItemClickListener

    internal constructor(
        listener: BookItemClickListener,
        bookList: MutableList<Book>?,
        isSearchType: Boolean
    ) {
        this.listener = listener
        this.bookList = bookList
        this.isSearchType = isSearchType
    }

    internal constructor(listener: BookItemClickListener, isSearchType: Boolean) {
        this.listener = listener
        this.isSearchType = isSearchType
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): BookViewHolder {
        val inflatedResource = if (isSearchType) R.layout.item_search_book else R.layout.item_book
        val view = LayoutInflater.from(parent.context).inflate(inflatedResource, parent, false)
        val viewHolder = BookViewHolder(view, isSearchType)
        view.setOnClickListener { selectedView: View? ->
            listener.onItemClicked(
                selectedView,
                viewHolder.adapterPosition
            )
        }
        return viewHolder
    }

    override fun onBindViewHolder(holder: BookViewHolder, position: Int) {
        holder.setData(bookList!![position])
    }

    override fun getItemCount(): Int {
        return if (bookList == null) 0 else bookList!!.size
    }

    fun updateItems(books: List<Book>?) {
        if (this.bookList == null) {
            this.bookList = mutableListOf()
        }

        this.bookList?.clear()
        if (books != null) {
            this.bookList?.addAll(books)
        }
        notifyDataSetChanged()
    }

    class BookViewHolder(itemView: View, private val isSearchType: Boolean) :
        RecyclerView.ViewHolder(itemView), OnItemClickListener {
        private var bookImageView: ImageView? = null
        private val bookNameTextView: TextView
        private val authorTextView: TextView
        private var editionTextView: TextView? = null
        private var descriptionTextView: TextView? = null

        init {
            if (isSearchType) {
                descriptionTextView = itemView.findViewById(R.id.book_description_text_view)
            } else {
                bookImageView = itemView.findViewById(R.id.book_image_view)
                editionTextView = itemView.findViewById(R.id.edition_text_view)
            }
            bookNameTextView = itemView.findViewById(R.id.book_name_text_view)
            authorTextView = itemView.findViewById(R.id.book_author_text_view)
        }

        fun setData(book: Book) {
            if (isSearchType) {
                descriptionTextView!!.text = book.description
            } else {
                //Set book
                editionTextView!!.text = String.format(Locale.ENGLISH, "Edition %d", book.edition)
            }
            bookNameTextView.text = book.name
            authorTextView.text = book.author
        }

        override fun onItemClick(adapterView: AdapterView<*>?, view: View, i: Int, l: Long) {
            Log.e("Selcted Item", "onItemClick: $i")
        }
    }
}